securesystemslib/signer/_azure_signer.py

```diff
@@ -0,0 +1,272 @@
+"""Signer implementation for Azure Key Vault"""
+
+import binascii
+import logging
+from typing import Optional, Tuple
+from urllib import parse
+
+import securesystemslib.hash as sslib_hash
+from securesystemslib.exceptions import UnsupportedLibraryError
+from securesystemslib.signer._key import Key
+from securesystemslib.signer._signer import (
+    SecretsHandler,
+    Signature,
+    Signer,
+    SSlibKey,
+)
+
+AZURE_IMPORT_ERROR = None
+try:
+    from azure.core.exceptions import HttpResponseError
+    from azure.identity import DefaultAzureCredential
+    from azure.keyvault.keys import KeyClient, KeyCurveName, KeyVaultKey
+    from azure.keyvault.keys.crypto import (
+        CryptographyClient,
+        SignatureAlgorithm,
+    )
+    from cryptography.hazmat.primitives.asymmetric import ec
+    from cryptography.hazmat.primitives.asymmetric.utils import (
+        encode_dss_signature,
+    )
+    from cryptography.hazmat.primitives.serialization import (
+        Encoding,
+        PublicFormat,
+    )
+except ImportError:
+    AZURE_IMPORT_ERROR = (
+        "azure-cloud-kms library required to sign with Azure Cloud keys."
+    )
+
+logger = logging.getLogger(__name__)
+
+
+class UnsupportedKeyType(Exception):
+    pass
+
+
+class AzureSigner(Signer):
+    """Azure Key Vault Signer
+
+    This Signer uses Azure Key Vault to sign.
+    Currently this signer only supports signing with EC keys.
+    RSA support will be added in a separate pull request.
+
+    Arguments:
+        az_key_uri: Fully qualified Azure Key Vault name, like
+            https://<vault-name>.vault.azure.net/keys/<key-name>/<version>
+        public_key: public key object
+
+    Raises:
+        Various errors from azure.identity
+        Various errors from azure.keyvault.keys
+    """
+
+    SCHEME = "azurekms"
+
+    def __init__(self, az_key_uri: str, public_key: Key):
+        if AZURE_IMPORT_ERROR:
+            raise UnsupportedLibraryError(AZURE_IMPORT_ERROR)
+
+        try:
+            cred = DefaultAzureCredential()
+            self.crypto_client = CryptographyClient(
+                az_key_uri,
+                credential=cred,
+            )
+            self.signature_algorithm = self._get_signature_algorithm(
+                public_key,
+            )
+            self.hash_algorithm = self._get_hash_algorithm(public_key)
+        except UnsupportedKeyType as e:
```

Comment:
Should UnsupportedKeyType be public API? Currently it isn't, so a user can't really handle it. Would a built-in exception type also do the trick?

Also, is it really worth to catch, log and re-raise? Doesn't the original error message, which you control, provide enough info? This pattern seems more useful below, where you contextualise the 3rd-party HttpResponseError, although even there I wonder if the stack trace alone might be enough. 🤷 

FYI: we have a discussion ticket about Signer exceptions in https://github.com/secure-systems-lab/securesystemslib/issues/468. Curious about your opinions.

Reply:
I left a note about this: https://github.com/secure-systems-lab/securesystemslib/pull/588#discussion_r1211249464

TLDR: I think it's not really public API, it's just part of the larger group of exceptions that we know we may raise that callers are not really meant to handle

Reply:
Agreed. I was actually trying to challenge the need for a custom exception and the way it is handled internally. But I'm okay with it. The code is still very readable and extra info for the caller is nice.